src/runtime/relax_vm/cuda/cuda_graph_builtin.cc

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/runtime/relax_vm/cuda_graph_builtin.cc
+ * \brief The CUDA graph related builtin functions for Relax virtual machine.
+ */
+
+#include <tvm/runtime/container/adt.h>
+#include <tvm/runtime/data_type.h>
+#include <tvm/runtime/device_api.h>
+#include <tvm/runtime/logging.h>
+#include <tvm/runtime/memory.h>
+#include <tvm/runtime/ndarray.h>
+#include <tvm/runtime/packed_func.h>
+#include <tvm/runtime/registry.h>
+#include <tvm/runtime/relax_vm/bytecode.h>
+#include <tvm/runtime/relax_vm/memory_manager.h>
+#include <tvm/runtime/relax_vm/vm.h>
+
+#include "../../cuda/cuda_common.h"
+namespace tvm {
+namespace runtime {
+namespace relax_vm {
+
+using tvm::runtime::NDArray;
+
+class CUDAGraphNode : public Object {
+ public:
+  cudaGraph_t handle_;
+  TVM_DECLARE_FINAL_OBJECT_INFO(CUDAGraphNode, Object);
+};
+
+class CUDAGraph : public ObjectRef {
+ public:
+  explicit CUDAGraph(cudaGraph_t handle) {
+    auto n = make_object<CUDAGraphNode>();
+    n->handle_ = handle;
+    data_ = std::move(n);
+  }
+  TVM_DEFINE_OBJECT_REF_METHODS(CUDAGraph, ObjectRef, CUDAGraphNode);
+};
+
+class CUDAGraphCache : public Object {
+ public:
+  struct Entry {
+    std::vector<ObjectRef> states;
+    CUDAGraph graph;
+  };
+  std::unordered_map<ObjectRef, Entry, ObjectPtrHash, ObjectPtrEqual> entries;
+
+  static CUDAGraphCache* Get() { return dmlc::ThreadLocalStore<CUDAGraphCache>::Get(); }
+};
+
+struct CaptureContext {
+  cudaGraph_t graph;
+  cudaStream_t stream;
+  // TODO: refactor CUDA graph rewrite pass so that warm up run is not needed
+  bool is_warm_up;
+};
+
+typedef dmlc::ThreadLocalStore<CaptureContext*> CaptureContextStore;
+
+TVM_REGISTER_GLOBAL("vm.builtin.get_captured_cuda_graph")
```

Comment:
Minor for later considration. consider pass in a global index integer (assigned during in rewrite) as key, think about how to speedup lookup time. Another thing to consider is that we may not get the same func address every time.

Reply:
agree using global index would be better. A scenario that we need to consider is multiple VM exist at the same time, where such global index could cause conflict. We will need to keep the cache per-VM

Reply:
Agree, we can start from this one for now